test.ts

```diff
@@ -142,8 +141,7 @@ test('cache option', t => {
 	const fixture = () => i++;
 	const memoized = mem(fixture, {
 		cache: new WeakMap(),
-		// @ts-expect-error Due to https://github.com/sindresorhus/mem/issues/50
-		cacheKey: ([firstArgument]) => firstArgument
+		cacheKey: ([firstArgument]: [any]) => firstArgument
```

Comment:
I don't think this is the right type and this test was essentially wrong. There was never a `firstArgument` so this shouldn't pass. The type of cacheKey should be

```ts
(...arguments: Parameters<fixture>) = CacheKeyType
```

`Parameters<fixture>` is strictly `[]`, not `any[]`.

Reply:
Live demo: https://www.typescriptlang.org/play?#code/DYUwLgBAlhC8EAYDcAoAxgewHYGdIDMoAPMAVwCcQ4IAKASjgD5oBqF1MATwAcqBhAIZoAFiADSITtRoA6OQPIBzUgFsQWMDgD6ALggAFBQLVgQ5HAB4uvDPgiESFEIwaxmArJ1SZckNENEJTj1BEXFJaVdmAGUwcigsRRoARjokCAB6DIgAeTEAQnRsPAh-MKCAJhCA8Kl4GgBtQnMwAEElVXUwAF0oiFj4xJpmvHblNQ00zOyAORyAFVyCoA

Reply:
I believe there are 2 flaws in what you're doing. First, in your TS playground example you have cacheKey receiving `(...arguments_: Parameters<typeof fixture>) => any;`. That isn't correct. cacheKey receives `(arguments_: Parameters<typeof fixture>) => any`. The first argument IS the array. The additional `...` causes it to flatten out. That was the flaw that was in `master` branch.

The other issue is that according to `mem` code we **always** have a firstArgument, because it's the array.

```
const memoized = function (this: any, ...arguments_) {
    const key = cacheKey ? cacheKey(arguments_) : arguments_[0];
```

In example in your js console if you run

```
function foo(...arguments_) { console.log(arguments_); }
console.log(foo())
// []
console.log(foo("bar"))
// ["bar"]
console.log(foo("bar", "baz"))
// ["bar", "baz"]
```

So then cacheKey will receive `[]`. In this circumstance we're effectively calling a function which we gave TS no type hinting. In the code of the test we're returning `String(firstArgument)` and that's totally valid because we're sometimes passing it an int, and sometimes a string, so we have to instruct TS that we know what we're doing and `any` it or `string | number`. The intent of this test isn't really to verify the TS integrity, it's to verify the caching, so I believe using `any` is totally valid.

Reply:
> The other issue is that according to `mem` code we **always** have a firstArgument, because it's the array.

That's slightly incorrect. It does have `arguments_`, not `[firstArgument]`. Notice that the latter is being destructured. `arguments_` will be the tuple `[]`, so there can't be a `[firstArgument]` in there.



> so I believe using `any` is totally valid.

For the reason I just explained, it's not. You can see this in the latest type tests I added, which are being ignored: https://github.com/sindresorhus/mem/commit/837e393531fa26002f860ff038901427597009fd

`any[]` and `[any]` can't be assigned to `[]`.